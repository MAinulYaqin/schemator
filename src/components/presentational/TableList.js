import React from "react";
import PropTypes from "prop-types";

import TableBox from "./TableBox";

const TableList = ({
  tables,
  fields,
  onMouseDown,
  onMouseUp,
  onMouseMove,
  onMouseEnter,
  onMouseLeave,
  onContextMenu,
  onClickAddField,
  onClickRemoveField,
  onChangeField,
  onChangeName,
  onChangeOptions
}) => (
  <g>
    {tables.map(table => {
      const byTableID = ({ tableID }) => tableID === table.id;
      const currentFields = fields.filter(byTableID);

      return (
        <TableBox
          key={table.id}
          ref={table.ref}
          {...table}
          fields={currentFields}
          options={table.options}
          onMouseDown={event => onMouseDown(event, table.id)}
          onMouseUp={event => onMouseUp(event, table.id)}
          onMouseMove={event => onMouseMove(event, table.id)}
          onMouseEnter={() => onMouseEnter(table.id)}
          onMouseLeave={() => onMouseLeave(table.id)}
          onContextMenu={() => onContextMenu(table.id)}
          onClickAddField={() => onClickAddField(table.id)}
          onClickRemoveField={onClickRemoveField}
          onChangeFieldName={(event, fieldID) =>
            onChangeField(event, fieldID, "name")
          }
          onChangeFieldType={(event, fieldID) =>
            onChangeField(event, fieldID, "type")
          }
          onChangeName={event => onChangeName(event, table.id)}
          onChangeOptions={(event, name) =>
            onChangeOptions(event, table.id, name)
          }
        />
      );
    })}
  </g>
);

TableList.propTypes = {
  tables: PropTypes.array,
  fields: PropTypes.array,
  onMouseDown: PropTypes.func,
  onMouseUp: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onContextMenu: PropTypes.func,
  onClickAddField: PropTypes.func,
  onClickRemoveField: PropTypes.func,
  onChangeField: PropTypes.func,
  onChangeName: PropTypes.func,
  onChangeOptions: PropTypes.func
};

export default TableList;
