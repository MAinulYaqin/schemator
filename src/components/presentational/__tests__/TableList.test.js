import React, { createRef } from "react";
import { shallow } from "enzyme";

import TableList from "../TableList";

jest.mock("../TableBox", () => "TableBox");

describe("TableList", () => {
  const setup = propOverrides => {
    const props = {
      tables: [
        {
          id: "a79c8abc-bcc6-40e4-9c52-b8f343b91fec",
          ref: createRef(),
          name: "User",
          timestamp: 1530376280298,
          position: {
            x: 92,
            y: 110
          },
          options: {
            id: true,
            rememberToken: true,
            softDeletes: false,
            timestamps: true
          }
        },
        {
          id: "1817b56b-d1ba-47f9-983e-9f9fdc90a533",
          ref: createRef(),
          name: "Post",
          timestamp: 1530377404366,
          position: {
            x: 548,
            y: 784
          },
          options: {
            id: true,
            rememberToken: false,
            softDeletes: false,
            timestamps: true
          }
        }
      ],
      fields: [
        {
          id: "079cff9a-7603-4d7a-b178-2da03d6faf8c",
          tableID: "a79c8abc-bcc6-40e4-9c52-b8f343b91fec",
          name: "name",
          type: "STRING"
        },
        {
          id: "34fd94df-d5d4-41ab-8f01-a2b95a2b2050",
          tableID: "a79c8abc-bcc6-40e4-9c52-b8f343b91fec",
          name: "email",
          type: "STRING"
        },
        {
          id: "6f159bff-ebe0-4d4e-8794-b75a1a34f79c",
          tableID: "a79c8abc-bcc6-40e4-9c52-b8f343b91fec",
          name: "password",
          type: "STRING"
        },
        {
          id: "26737c00-7a54-4b58-b112-f17060fa49f6",
          tableID: "1817b56b-d1ba-47f9-983e-9f9fdc90a533",
          name: "title",
          type: "INTEGER"
        },
        {
          tableID: "1817b56b-d1ba-47f9-983e-9f9fdc90a533",
          id: "ccbc8a47-ae75-4ab6-8d6f-7a90747d60be",
          name: "user_id",
          type: "INTEGER"
        },
        {
          tableID: "1817b56b-d1ba-47f9-983e-9f9fdc90a533",
          id: "5c0b69d3-f527-4a02-a095-1342f392b2c8",
          name: "content",
          type: "TEXT"
        }
      ],
      onMouseDown: jest.fn(),
      onMouseUp: jest.fn(),
      onMouseMove: jest.fn(),
      onMouseEnter: jest.fn(),
      onMouseLeave: jest.fn(),
      onContextMenu: jest.fn(),
      onClickAddField: jest.fn(),
      onClickRemoveField: jest.fn(),
      onChangeField: jest.fn(),
      onChangeName: jest.fn(),
      onChangeOptions: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<TableList {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("should calls 'onMouseDown' while mouse down on TableBox", () => {
    const { wrapper, props } = setup();
    const event = {};

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onMouseDown(event, table.id);

      expect(props.onMouseDown).toBeCalledWith(event, table.id);
    });
  });

  it("should calls 'onMouseUp' while mouse up on TableBox", () => {
    const { wrapper, props } = setup();
    const event = {};

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onMouseUp(event, table.id);

      expect(props.onMouseUp).toBeCalledWith(event, table.id);
    });
  });

  it("should calls 'onMouseMove' while mouse move on TableBox", () => {
    const { wrapper, props } = setup();
    const event = {};

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onMouseMove(event, table.id);

      expect(props.onMouseMove).toBeCalledWith(event, table.id);
    });
  });

  it("should calls 'onMouseEnter' while mouse entering TableBox", () => {
    const { wrapper, props } = setup();

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onMouseEnter(table.id);

      expect(props.onMouseEnter).toBeCalledWith(table.id);
    });
  });

  it("should calls 'onMouseLeave' while mouse leaving TableBox", () => {
    const { wrapper, props } = setup();

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onMouseLeave(table.id);

      expect(props.onMouseLeave).toBeCalledWith(table.id);
    });
  });

  it("should calls 'onContextMenu' while context menu shown in TableBox", () => {
    const { wrapper, props } = setup();

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onContextMenu(table.id);

      expect(props.onContextMenu).toBeCalledWith(table.id);
    });
  });

  it("should calls 'onClickAddField' while add field button clicked in TableBox", () => {
    const { wrapper, props } = setup();

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onClickAddField(table.id);

      expect(props.onClickAddField).toBeCalledWith(table.id);
    });
  });

  it("should calls 'onClickRemoveField' while remove field button clicked in TableBox", () => {
    const { wrapper, props } = setup();

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onClickRemoveField();

      expect(props.onClickRemoveField).toBeCalled();
    });
  });

  it("should calls 'onChangeField' while field name changed in TableBox", () => {
    const { wrapper, props } = setup();
    const event = {};
    const fieldID = "field-id-12345";

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onChangeFieldName(event, fieldID);

      expect(props.onChangeField).toBeCalledWith(event, fieldID, "name");
    });
  });

  it("should calls 'onChangeField' while field type changed in TableBox", () => {
    const { wrapper, props } = setup();
    const event = {};
    const fieldID = "field-id-12345";

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onChangeFieldType(event, fieldID);

      expect(props.onChangeField).toBeCalledWith(event, fieldID, "type");
    });
  });

  it("should calls 'onChangeName' while table name changed in TableBox", () => {
    const { wrapper, props } = setup();
    const event = {};

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onChangeName(event, table.id);

      expect(props.onChangeName).toBeCalledWith(event, table.id);
    });
  });

  it("should calls 'onChangeOptions' while options changed in TableBox", () => {
    const { wrapper, props } = setup();
    const event = {};
    const name = "rememberToken";

    props.tables.forEach((table, index) => {
      wrapper
        .find("TableBox")
        .at(index)
        .props()
        .onChangeOptions(event, name);

      expect(props.onChangeOptions).toBeCalledWith(event, table.id, name);
    });
  });
});
