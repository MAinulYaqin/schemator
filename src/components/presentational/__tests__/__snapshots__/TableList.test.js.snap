// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableList should renders default 1`] = `
<g>
  <TableBox
    fields={
      Array [
        Object {
          "id": "079cff9a-7603-4d7a-b178-2da03d6faf8c",
          "name": "name",
          "tableID": "a79c8abc-bcc6-40e4-9c52-b8f343b91fec",
          "type": "STRING",
        },
        Object {
          "id": "34fd94df-d5d4-41ab-8f01-a2b95a2b2050",
          "name": "email",
          "tableID": "a79c8abc-bcc6-40e4-9c52-b8f343b91fec",
          "type": "STRING",
        },
        Object {
          "id": "6f159bff-ebe0-4d4e-8794-b75a1a34f79c",
          "name": "password",
          "tableID": "a79c8abc-bcc6-40e4-9c52-b8f343b91fec",
          "type": "STRING",
        },
      ]
    }
    id="a79c8abc-bcc6-40e4-9c52-b8f343b91fec"
    key="a79c8abc-bcc6-40e4-9c52-b8f343b91fec"
    name="User"
    onChangeFieldName={[Function]}
    onChangeFieldType={[Function]}
    onChangeName={[Function]}
    onChangeOptions={[Function]}
    onClickAddField={[Function]}
    onClickRemoveField={[Function]}
    onContextMenu={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    options={
      Array [
        Object {
          "id": "id",
          "isChecked": true,
          "label": "ID",
        },
        Object {
          "id": "rememberToken",
          "isChecked": false,
          "label": "Remember Token",
        },
        Object {
          "id": "softDeletes",
          "isChecked": false,
          "label": "Soft Deletes",
        },
        Object {
          "id": "timestamps",
          "isChecked": false,
          "label": "Timestamps",
        },
      ]
    }
    position={
      Object {
        "x": 92,
        "y": 110,
      }
    }
    timestamp={1530376280298}
    types={
      Array [
        "INTEGER",
        "STRING",
        "TEXT",
      ]
    }
  />
  <TableBox
    fields={
      Array [
        Object {
          "id": "26737c00-7a54-4b58-b112-f17060fa49f6",
          "name": "title",
          "tableID": "1817b56b-d1ba-47f9-983e-9f9fdc90a533",
          "type": "INTEGER",
        },
        Object {
          "id": "ccbc8a47-ae75-4ab6-8d6f-7a90747d60be",
          "name": "user_id",
          "tableID": "1817b56b-d1ba-47f9-983e-9f9fdc90a533",
          "type": "INTEGER",
        },
        Object {
          "id": "5c0b69d3-f527-4a02-a095-1342f392b2c8",
          "name": "content",
          "tableID": "1817b56b-d1ba-47f9-983e-9f9fdc90a533",
          "type": "TEXT",
        },
      ]
    }
    id="1817b56b-d1ba-47f9-983e-9f9fdc90a533"
    key="1817b56b-d1ba-47f9-983e-9f9fdc90a533"
    name="Post"
    onChangeFieldName={[Function]}
    onChangeFieldType={[Function]}
    onChangeName={[Function]}
    onChangeOptions={[Function]}
    onClickAddField={[Function]}
    onClickRemoveField={[Function]}
    onContextMenu={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    options={
      Array [
        Object {
          "id": "id",
          "isChecked": true,
          "label": "ID",
        },
        Object {
          "id": "rememberToken",
          "isChecked": false,
          "label": "Remember Token",
        },
        Object {
          "id": "softDeletes",
          "isChecked": false,
          "label": "Soft Deletes",
        },
        Object {
          "id": "timestamps",
          "isChecked": false,
          "label": "Timestamps",
        },
      ]
    }
    position={
      Object {
        "x": 548,
        "y": 784,
      }
    }
    timestamp={1530377404366}
    types={
      Array [
        "INTEGER",
        "STRING",
        "TEXT",
      ]
    }
  />
</g>
`;
